test/src/test/java/jenkins/model/NodesTest.java

```diff
@@ -168,6 +171,64 @@ public void replaceNodeShouldNotRemoveIdenticalOldNode() throws Exception {
         assertNotNull(r.jenkins.getNode("foo"));
     }
 
+    @Test
+    @Issue("JENKINS-33704")
+    public void replacingSecondNodeIsLocal() throws Exception {
+        DumbSlave nodeA = r.createSlave("nodeA", "temp", null);
+        var retentionStrategyA = new MockRetentionStrategy();
+        nodeA.setRetentionStrategy(retentionStrategyA);
+        r.jenkins.addNode(nodeA);
+        assertThat(retentionStrategyA.checkCount, equalTo(0));
+        DumbSlave nodeB = r.createSlave("nodeB", "temp", null);
+        var retentionStrategyB = new MockRetentionStrategy();
+        nodeB.setRetentionStrategy(retentionStrategyB);
+        r.jenkins.addNode(nodeB);
+        assertThat(retentionStrategyA.checkCount, equalTo(0));
+        assertThat(retentionStrategyB.checkCount, equalTo(0));
+        Jenkins.get().getNodesObject().replaceNode(nodeA, nodeA);
+        assertThat(retentionStrategyB.checkCount, equalTo(0));
+    }
+
+    @Test
+    @Issue("JENKINS-33704")
+    public void removingSecondNodeIsLocal() throws Exception {
+        DumbSlave nodeA = r.createSlave("nodeA", "temp", null);
+        var retentionStrategyA = new MockRetentionStrategy();
+        nodeA.setRetentionStrategy(retentionStrategyA);
+        r.jenkins.addNode(nodeA);
+        assertThat(retentionStrategyA.checkCount, equalTo(0));
+        DumbSlave nodeB = r.createSlave("nodeB", "temp", null);
+        var retentionStrategyB = new MockRetentionStrategy();
+        nodeB.setRetentionStrategy(retentionStrategyB);
+        r.jenkins.addNode(nodeB);
+        assertThat(retentionStrategyA.checkCount, equalTo(0));
+        assertThat(retentionStrategyB.checkCount, equalTo(0));
+        r.jenkins.removeNode(nodeA);
+        assertThat(retentionStrategyB.checkCount, equalTo(0));
+    }
+
+    @Test
+    @Issue("JENKINS-33704")
+    public void changingBuiltInNodeDoesntChangeOtherNodes() throws Exception {
+        DumbSlave nodeA = r.createSlave("nodeA", "temp", null);
+        var retentionStrategyA = new MockRetentionStrategy();
+        nodeA.setRetentionStrategy(retentionStrategyA);
+        r.jenkins.addNode(nodeA);
+        assertThat(retentionStrategyA.checkCount, equalTo(0));
+        r.jenkins.setNumExecutors(1);
+        assertThat(retentionStrategyA.checkCount, equalTo(0));
+    }
+
+    public static class MockRetentionStrategy extends RetentionStrategy.Always {
+        private int checkCount = 0;
```

Comment:
Would be helpful for reviewers to note at which points in the test cases this count is positive.

Reply:
Also, is it intentional that none of the tests expect a positive count? When would `check` now be called in the course of updating the node list?

Reply:
Expected callers are the cron-based retention strategy check (https://github.com/jenkinsci/jenkins/blob/16b5d210525f1e0b31d15b6cf4a06c291099e049/core/src/main/java/hudson/slaves/ComputerRetentionWork.java#L78), and https://github.com/jenkinsci/jenkins/blob/ba900e9acf01fe963716e49e200c9f027bf63c19/core/src/main/java/hudson/slaves/SlaveComputer.java#L972 that is called whenever the node gets updated (but not when initially added)